### PR TITLE
Add vision anomaly detection pipeline with PaSTe model

### PR DESCRIPTION
**Pull request by LukaszMarcjan**
## Summary

Adds vision anomaly detection support to pyCLAD. PaSTe is the first vision model, implemented in PyTorch Lightning, with a configurable backbone (e.g. `mobilenet_v2`, `resnet18`, `wide_resnet50_2`).

To add another vision model, implement the training logic and a `score_maps()` method. A model opts into pixel-level evaluation just by exposing `score_maps()`; if it's not there, the callback is a no-op (no config flags needed).

The generic folder reader handles five layout variants through shared indexing logic. Default dataset root is `examples/resources/vision/`, which keeps downloaded benchmarks inside the project tree; override with `PYCLAD_VISION_DATASETS_ROOT` or the registry file.

Additional features:
- Pixel-level evaluation on top of the existing `ConceptMetricCallback`. `VisionPixelConceptMetricCallback` pulls score maps from the model and compares them against ground-truth masks.
- Dataset loading for five benchmarks: MVTec AD, BTech, DAGM, MPDD, VisA,and  arbitrary custom folders with automatic layout detection. Datasets are resolved through a registry: explicit paths, env vars, or a shared JSON.
- HuggingFace integration for downloading prebuilt benchmark manifests with reproducible category orderings.

## Testing

Unit tests cover: dataset readers and registry resolution, manifest round-trips and category ordering, pixel-level metric edge cases, the callback with and without masks, and base metric computation.

End-to-end run on MVTec AD: 150 train / 150 test images per category, 5 epochs. The other benchmarks (BTech, DAGM, MPDD, VisA) got 15 images per category and 5 epochs — mostly a smoke test to confirm data loading and the pipeline hold together.

### CODE REVIEW

File: src/pyclad/vision/data/base.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+                    VisionConcept(
+                        name=category,
+                        data=materialize_samples(
+                            samples=test_samples,
+                            data_mode=data_mode,
+                            resize_to=resize_to,
+                            color_mode=color_mode,
+                        ),
+                        labels=np.asarray([s.image_label for s in test_samples], dtype=np.int64),
+                        masks=masks,
```
Comment (Nyderx) [inline src/pyclad/vision/data/base.py:126]: We don't need to pass `kept_indices` here to make sure that we do not create misaligment? Is it possible that `VisionConcept.masks` will have less rows than `VisionConcept.data` and `VisionConcept.labels` if we skip some anomalies with `skip_missing_anomaly_masks=True` in load_ground_truth_masks_for_samples?

File: src/pyclad/data/vision/base.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+                    color_mode=color_mode,
+                ),
+                labels=None,
+            )
+        )
+
+        if len(test_samples) > 0:
+            has_any_mask = any(s.mask_path is not None for s in test_samples)
+            if has_any_mask:
+                from pyclad.data.vision.masks import load_ground_truth_masks_for_samples
```
Comment (Nyderx) [inline src/pyclad/data/vision/base.py]: Maybe let's keep all imports at the beginning of the file unless they bring significant additional module loading time?

File: src/pyclad/models/vision/paste/architecture.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        anomaly_map = aggregated_map[:, 0]
+        anomaly_scores = torch.max(anomaly_map.view(anomaly_map.size(0), -1), dim=1).values
+        return anomaly_map, anomaly_scores
+
+    @staticmethod
+    def feature_loss(teacher_features: list[torch.Tensor], student_features: list[torch.Tensor]) -> torch.Tensor:
+        if len(teacher_features) == 0 or len(student_features) == 0:
+            raise ValueError("PaSTe feature loss requires non-empty teacher and student feature lists")
+
+        loss = torch.sum((teacher_features[0] - student_features[0]) ** 2, dim=1).mean()
```
Comment (Nyderx) [inline src/pyclad/models/vision/paste/architecture.py]: I've got this concern from automated review process. Could you please verify?
```
  [Major] Feature loss applies normalization inconsistently across layers
  models/vision/paste/architecture.py:121-130
  
  # First feature pair: raw MSE — no normalization
  loss = torch.sum((teacher_features[0] - student_features[0]) ** 2, dim=1).mean()
  # All subsequent pairs: normalized
  for teacher_feature, student_feature in zip(teacher_features[1:], student_features[1:]):
      teacher_feature = F.normalize(teacher_feature, dim=1)
      student_feature = F.normalize(student_feature, dim=1)
      loss = loss + ...

  But during inference in post_process, all feature pairs (including the first) are normalized. The training loss and the inference distance metric are measuring different feature spaces for the first AD layer. In practice the first layer's raw scale will dominate the gradient, making subsequent layers
  contribute disproportionately less to training while contributing equally during scoring. This should likely be uniform normalization across all layers in both training and inference.
```

File: src/pyclad/data/vision/vision_reader.py
Comment (Nyderx) [inline src/pyclad/data/vision/vision_reader.py:1]: Compare with `vision/base.py` and decide which one we want to keep (or explain why we need both)

File: src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                cb._learned_concepts = original_concepts
+
+            for key, value in cb_info.items():
+                pixel_key = key.replace("concept_metric_callback_", "pixel_concept_metric_callback_")
+                value["evaluation_level"] = "pixel"
+                result[pixel_key] = value
+        return result
+
+
+def _resolve_model(strategy: Any, concept_name: str):
```
Comment (Nyderx) [inline src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py]: This seems to be very error prone (what if strategy does not have `_models`?). It means using a reflection to access a specific private part of the model, breaking the contract. We should deal with this in a different way (for example adding getter function in the contract, providing more data from the model `predict`, etc)

File: src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        if not self._has_evaluations:
+            return {}
+        result: Dict[str, Any] = {}
+        for cb in self._callbacks:
+            concepts_with_masks = set()
+            for evals in cb._metric_matrix.values():
+                concepts_with_masks.update(evals.keys())
+
+            original_concepts = cb._learned_concepts
+            cb._learned_concepts = [c for c in original_concepts if c in concepts_with_masks]
```
Comment (Nyderx) [inline src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py]: Why are we mutating things in the `info` function?

File: src/pyclad/models/vision/utilities/backbones.py
Comment (Nyderx) [inline src/pyclad/models/vision/utilities/backbones.py:1]: Shouldn't we use things from here in `paste`?

File: src/pyclad/data/vision/benchmarks/registry.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    VisionBenchmarkSpec,
+    build_vision_benchmark_reader,
+    index_vision_benchmark,
+    read_vision_benchmark_dataset,
+)
+
+DEFAULT_VISION_DATASET_REGISTRY_PATH = (
+    Path(__file__).resolve().parents[2] / "datasets" / "vision_datasets" / "registry.json"
+)
+DEFAULT_VISION_DATASETS_ROOT = Path(__file__).resolve().parents[5] / "examples" / "resources" / "vision"
```
Comment (Nyderx) [inline src/pyclad/data/vision/benchmarks/registry.py]: Let's use Paths from `paths`.  Doing `parent[x]` is error prone in this case.

File: src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py (hunk 4 above)
Comment (Nyderx) [reply src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py]: Maybe extending `Model` to `PixelLevelModel` or something like that?

**Review by Nyderx — CHANGES_REQUESTED** [review]
Thank you for doing the pull request and contributing to the library :) I added a few comments

File: src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py (hunk 4 above)
Comment (Nyderx) [reply src/pyclad/callbacks/evaluation/vision_pixel_concept_metric_callback.py]: This does not seem to be fixed 
image

File: src/pyclad/strategies/strategy.py [hunk 7]
```diff
@@ -18,6 +19,20 @@ def additional_info(self) -> Dict:
     def info(self) -> Dict[str, Any]:
         return {"strategy": {"name": self.name(), **self.additional_info()}}
 
+    def model_for_concept(self, concept_id: str) -> Optional[Model]:
```
Comment (Nyderx) [inline src/pyclad/strategies/strategy.py]: It's an improvement, but not a solution. 

1. It could break in concept incremental and concept agnostic setup (as there is no `concept_id` provided there. ConceptIncrementalStrategy and ConceptAgnosticStrategy are explicitly defined to not carry a concept identity. Their learn() and predict() take only data — there's no concept_id parameter anywhere in their contract. Putting a  method model_for_concept(concept_id: str) on a strategy type that, by definition, doesn't reason about concept IDs is a contract violation. 
2. the way it is designed also means that anything that uses model_for_concept would break in concept incremental and concept agnostic
3. Strategy is a behavioral abstraction — it defines learn(), predict(), and name(). It says nothing about how a strategy is implemented internally. Adding model_for_concept leaks an   implementation assumption (that every strategy is backed by one or more Model objects) into the core contract.
4. `return getattr(self, "_model", None)` is still a fragile reflection that was in the callback before the fix. It wasn't eliminated — it was relocated from the callback into strategy.py.

Comment (Nyderx) [reply src/pyclad/strategies/strategy.py]: We need it only for `score_maps`, right? 

I tried to prepare a suggestion using claude code to work across the codebase.  You can see diff here: https://github.com/lifelonglab/pyCLAD/commit/a026e5acda5d9ef2dc900db888e88a715d147d5b
and code is available on `paste-adjusted-interface` branch.

I didn't have time to analyze it thoroughly and test, so please do it. If it works and makes sense, we can add the commit to this PR. 

**Solution description:**
 Model.predict() now returns a PredictionResults dataclass instead of a tuple:
  
  @dataclass
  class PredictionResults:
      y_pred: np.ndarray
      anomaly_scores: np.ndarray

  @dataclass
  class VisionPredictionResults(PredictionResults):
      score_maps: np.ndarray

  Vision models override predict() with a covariant return type and compute everything — predictions, image-level scores, and pixel maps — in a single forward pass. Because strategies like NaiveStrategy simply delegate to the model (return self._model.predict(data)), they automatically propagate VisionPredictionResults when wrapping a vision model — no changes needed to the strategy. The  same strategy class works for both tabular and vision.

**Why is it better?** 
- No reflection. getattr(self, "_model", None) and isinstance(strategy, MSTE) are gone. The result carries what it carries; no strategy internals are inspected.
  - One forward pass. Previously predict() and score_maps() each looped over batches independently. Now one call to predict() returns everything.
  - Same strategy for both scenarios. A NaiveStrategy(PaSTe()) works in tabular and vision contexts without subclassing or adding methods. The richer result flows through automatically.
  - Vision concerns stay out of core abstractions. Strategy, Model, and all three base scenarios have no vision imports. The distinction lives in the result type, not the infrastructure.
  - Data flow is explicit and traceable. score_maps travels as a field on the return value through predict() → scenario → after_evaluation kwarg. Previously it was a side-channel retrieved by the scenario reaching into the
  strategy's model.
  - Extensible by default. A future ExplainabilityResults(PredictionResults) with attribution maps follows the exact same pattern — no new base class methods, no new scenario variants.

**Review by Nyderx — CHANGES_REQUESTED** [review]

**Review by Nyderx — APPROVED** [review]
